torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,172 @@ def cached():
+    parameters or buffers of a parametrized :class:`torch.nn.Module`.
 
-    .. note ::
+    It is the type of ``module.parametrizations[tensor_name]`` when ``module[tensor_name]``
+    has been parametrized with :func:`register_parametrization`.
+
+    If the first registered parmetrization has a ``right_inverse`` that returns one tensor or
+    does not have a ``right_inverse`` (in which case we assume that ``right_inverse`` is the identity),
+    it will hold the tensor under the name ``original``.
+    If it has a ``right_inverse`` that returns more than one tensor, these will be registered as
+    ``original0``, ``original1``, ...
+
+    .. warning::
         This class is used internally by :func:`register_parametrization`. It is documented
         here for completeness. It should not be instantiated by the user.
 
     Args:
-        modules (iterable): an iterable of modules representing the parametrizations
+        modules (sequence): an sequence of modules representing the parametrizations
         original (Parameter or Tensor): parameter or buffer that is parametrized
     """
     original: Tensor
 
     def __init__(
-        self, modules: Iterable[Module], original: Union[Tensor, Parameter]
+        self, modules: Sequence[Module], original: Union[Tensor, Parameter]
     ) -> None:
+        # We require this because we need to treat differently the first parametrization
+        # This should never throw, unless this class is used from the outside
+        if len(modules) == 0:
+            raise ValueError("ParametrizationList requires one or more modules.")
+
         super().__init__(modules)
-        if isinstance(original, Parameter):
-            self.register_parameter("original", original)
-        else:
-            self.register_buffer("original", original)
 
-    def set_original_(self, value: Tensor) -> None:
-        r"""This method is called when assigning to a parametrized tensor.
+        # We check that the following invariants hold:
+        #    X = module.weight
+        #    Y = param.right_inverse(X)
+        #    assert isinstance(Y, Tensor) or isinstance(Y, collections.Sequence)
+        #    Z = param(Y) if isisntance(Y, Tensor) else param(*Y)
+        #    if isinstance(Y, collections.Sequence):
+        #       assert all(isinstance(t, Tensor) for t in Y)
+        #    # Consistency checks
+        #    assert X.dtype == Z.dtype and X.shape == Z.shape
+        #    # If it has one input, this allows to be able to use set_ to be able to
+        #    # move data to/from the original tensor without changing its id (which is what the
+        #    # optimiser uses to track parameters)
+        #    # We also require the shape to be the same to avoid the user shooting themselves
+        #    # in the foot. This second requirement may be dropped in the future
+        #    if isinstance(Y, Tensor)
+        #      assert X.dtype == Y.dtype and X.shape == Y.shape
```

Comment:
I don't think we want to enforce the shape equality for X and Y. In particular, set_ will work just fine with another Tensor of a different size.

Reply:
It is not necessary, but I thought that, in the same way that we disallow to change the `dtype` for users not to encounter weird behaviours, we should also disallow changing the shape. That being said, I can remove this constraint.

Reply:
On a second thought, let's not be too conservative. I'm removing this constraint, as I could imagine it being useful.